### ISSUE TITLE
marcel like pipes

### ISSUE THREAD

**Issue by jnoortheen** [xonsh/xonsh#3967]
I've known [marcel](https://github.com/geophile/marcel) for sometime and it is cool. Xonsh could get some inspiration from the project especially composing output from subprocess with python functions through pipe. It feels like doing functional programming and it is useful when filtering/formatting output.

Also there is this another project that uses pipes to pass values too. https://github.com/terrycojones/daudin

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by scopatz** [comment]
I am open to ideas here! Are there specific features that you would like to see implemented?

**Comment by jnoortheen** [comment]
Mainly I wanted to have the pipeline chaining as the means to compose functions. I achieve this with [pipe](https://github.com/JulienPalard/Pipe)

![image](https://user-images.githubusercontent.com/6702219/98630444-2f661e00-2341-11eb-9ad1-7be4ed3d155e.png)

But extending this to commands don't work
![image](https://user-images.githubusercontent.com/6702219/98630605-84099900-2341-11eb-8ccf-81d7691ca006.png)

1. So a pipeline support at language level to pass the value between commands and python functions. This can be used to replace usage of grep/awk/sed and all such filtering/formatting tools with python code. 
2. shorter lambda like `x: x.upper()` instead of `lambda x: x.upper()` . It is just a syntactic sugar for regular lambda. But less words to type. When working interactively on terminal, this can help together with piped functions as callbacks. But I think it is very less important, as this adds a new syntax and one needs to explain what is what.

**Comment by scopatz** [comment]
Interesting.  Would it be OK if you had to wrap the subcommand? That is `as_list | !(head 1)`.  Something like that could be achieved by giving the CommandPipeline object the right API.

**Comment by jnoortheen** [comment]
> Interesting. Would it be OK if you had to wrap the subcommand? That is `as_list | !(head 1)`. Something like that could be achieved by giving the CommandPipeline object the right API.

Yes I think that can be a good start for composing them together.

**Comment by scopatz** [comment]
We could do something like with the subprocess mode, where there is a formal version of the lang that includes the explicit mode, ie `as_list | !(head 1)`, but if something fails to parse, then xonsh would attempt to autowrap the expressions like `as_list | head 1` into `as_list | !(head 1)`. This might be tricky to jointly implement with the autowrapping of subprocesses, but if someone got this working it would be really cool

**Comment by anki-code** [comment]
[xontrib-pipeliner](https://github.com/anki-code/xontrib-pipeliner) -  Let your pipe lines flow thru the Python code in xonsh.

**Comment by deeuu** [comment]
@jnoortheen Still interested in this? I'd be happy to collaborate on such a feature!

**Comment by jnoortheen** [comment]
> @jnoortheen Still interested in this? I'd be happy to collaborate on such a feature!

Yeah sure. I still want to see such feature implemented. I think it will be best to have this work as other functional language's pipe operator `|>` (as in F#, elm, [coconut-pipeline](https://coconut.readthedocs.io/en/master/DOCS.html#id107)). But currently I am OK with [pipe](https://github.com/JulienPalard/Pipe). I am currently playing with the idea of 'fzf for xonsh's auto-completion ' and will not be working on this anytime soon. Please feel free take on this.

**Comment by deeuu** [comment]
> 'fzf for xonsh's auto-completion

Would you mind pointing me to your work on this? I also have a small fuzzy-tab completer `xontrib`  (just routes optional completions to a user-defined fuzzy-finder for selection) but haven't made it public yet as I was hoping to get it working with the `readline` shell. 

Happy to chat on the gitter channel if that's more appropriate.

Cheers

**Comment by jnoortheen** [comment]
>
> Would you mind pointing me to your work on this? I also have a small fuzzy-tab completer xontrib (just routes optional  completions to a user-defined fuzzy-finder for selection) but haven't made it public yet as I was hoping to get it working with the readline shell.

Yes sure we can talk this on gitter.

**Comment by anki-code** [comment]
I believe this can be implemented as a xontrib and this issue could be closed.

**Comment by jnoortheen** [comment]
I like it to be a language feature. @anki-code please check out the said projects. They use lambda functions as a primary way to interact with the output. I locally use 'pipe' package to import the said variables. But it would be cool to have richer piping experience

**Comment by jnoortheen** [comment]
@anki-code for example check this out, https://www.marceltheshell.org/list-processes 

This provides rich interaction with the output.
